Add module to calculate near field and radiation field on user-define…

HaoPengYY opened the pull request:
Add module to calculate electrical field on user-defined observation plane.

Commit:
Add module to calculate near field and radiation field on user-defined observation plane!

PrometheusPi commented:
Hi @HaoPengYY, 
thanks for providing an update to Clara2 via pull request. I will review it as soon as possible. 
Could you briefly describe, how I can execute your code and what I should expect when testing.

HaoPengYY commented:
@PrometheusPi Hi. Actually, I have performed the benchmarks described in Sec. 5.1 of your [thesis](https://zenodo.org/record/843510#.YMBfKb7iuUk), and I think they are enough to verify the feasibility of this module. I am now preparing a markdown to explain the code and the benchmarks. I don't know if I can put it here, or I can send it to you by email.

PrometheusPi commented:
@HaoPengYY Cool you performed these benchmarks. I agree, they are perfectly suited for testing this module. 
I would suggest that you upload the markdown file you are currently creating as a second commit to this pull request. (Just create a `/doc` directory in the clara2 git repo and put the file in there.) This would allow others to read your explanation as well.

Thanks a lot for your effort to provide this new module and its documentation. :+1:

HaoPengYY committed:
Add introduction and benchmarks to the new module

Introductions and benchmarks of the new module, which is devoted to calculate the near and far field of the electrons based on their trajectories, are included a file in /doc/Intro&Benchmark/readme.md .

HaoPengYY commented:
@PrometheusPi Hi. It is my first time to use git, I don't know if this is the right way. I put the markdown here as well.
[Intro&BenchmarkToUOP.zip](https://github.com/ComputationalRadiationPhysics/clara2/files/6660135/Intro.BenchmarkToUOP.zip)
Thanks to your code, I am now able to do a work I am interested in. But it will be a pain to run the code when a large number of trajectories are used as input files, maybe we are able to add them together before we dump the data files separately.

PrometheusPi commented:
Dear @HaoPengYY,
I am very sorry that I had no time so far to look into your pull request. Hopefully there will be more time next week. 
Please excuse the delayed review.

PrometheusPi commented:
I am very sorry @HaoPengYY for not having the time yet to look into your pull request. I will hopefully be able to do so in September. Please excuse the delay. I hope it is fine for you to work on your branch.

HaoPengYY commented:
@PrometheusPi  Please don't be worry. The code now work fine for us, we have done some work based on the code. And also I have done some updates, like adding the module which calculates the far-field radiation field, changing the mpi logic (worker processors are assigned a new job after finishing the last job), adding all spectrum or electric fields coherently during calculation (no need for post-processing, save a lot of memory and facilitate simulations with a large number of trajectories).  I would start a new pull request if everything is ready.
